Nil pointer exception when getting notification on newly initialized leader

It appears a race condition on the leader initialization and receiving a `GetNotifications()` call from clients.

```
Jan 17 00:40:16.942 |INF| Received NewTerm request component='internal-rpc-server' peer='10.44.2.121:48644' req={ namespace='default' shardId='12' term='18' }
Jan 17 00:40:16.992 |INF| [JOB 1] WAL file /data/db/default/shard-12/000020.log with log number 000020 stopped reading at offset: 0; replayed 0 keys in 0 batches component='pebble' shard=12
Jan 17 00:40:17.004 |INF| Created leader controller component='leader-controller' namespace='default' shard=12 term=17
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x19e484a]
goroutine 109 [running]:
github.com/streamnative/oxia/server.(*leaderController).dispatchNotifications(0xc000256200, {0x22b0308, 0xc000690690}, 0xc00038e100, {0x22b61b0, 0xc000514200})
	/src/oxia/server/leader_controller.go:807 +0x16a
github.com/streamnative/oxia/server.(*leaderController).GetNotifications.func1()
	/src/oxia/server/leader_controller.go:773 +0x4f
github.com/streamnative/oxia/common.DoWithLabels.func1({0x22b02d0?, 0xc000908c00?})
	/src/oxia/common/pprof.go:46 +0x13
runtime/pprof.Do({0x22b0308?, 0xc000690690?}, {{0xc0003197a0?, 0x18d5980?, 0xc00067ce28?}}, 0xc00067cf30)
	/usr/local/go/src/runtime/pprof/runtime.go:51 +0x9d
github.com/streamnative/oxia/common.DoWithLabels({0x22b0308, 0xc000690690}, 0xc000908b70, 0xc00038e300)
	/src/oxia/common/pprof.go:42 +0x376
created by github.com/streamnative/oxia/server.(*leaderController).GetNotifications in goroutine 107
	/src/oxia/server/leader_controller.go:765 +0x2f0
```

--- a/server/leader_controller.go
+++ b/server/leader_controller.go
@@ -793,6 +793,9 @@ func (lc *leaderController) GetNotifications(req *proto.NotificationsRequest, st
 		cancel()
 		return lc.ctx.Err()
 
+	case <-ctx.Done():
+		return ctx.Err()
+
 	case <-stream.Context().Done():
 		// The stream is getting closed
 		cancel()
@@ -810,7 +813,14 @@ func (lc *leaderController) dispatchNotifications(ctx context.Context, req *prot
 	if req.StartOffsetExclusive != nil {
 		offsetInclusive = *req.StartOffsetExclusive + 1
 	} else {
-		commitOffset := lc.quorumAckTracker.CommitOffset()
+		lc.Lock()
+		qat := lc.quorumAckTracker
+		lc.Unlock()
+
+		if qat == nil {
+			return errors.New("leader is not yet ready")
+		}
+		commitOffset := qat.CommitOffset()
 
 		// The client is creating a new notification stream and wants to receive the notification from the next
 		// entry that will be written.

--- a/server/leader_controller.go
+++ b/server/leader_controller.go
@@ -843,6 +843,11 @@ func (lc *leaderController) dispatchNotifications(ctx context.Context, req *prot
 		offsetInclusive = commitOffset + 1
 	}
 
+	return lc.iterateOverNotifications(ctx, stream, offsetInclusive)
+}
+
+func (lc *leaderController) iterateOverNotifications(ctx context.Context, stream proto.OxiaClient_GetNotificationsServer, startOffsetInclusive int64) error {
+	offsetInclusive := startOffsetInclusive
 	for ctx.Err() == nil {
 		notifications, err := lc.db.ReadNextNotifications(ctx, offsetInclusive)
 		if err != nil {

--- a/server/leader_controller_test.go
+++ b/server/leader_controller_test.go
@@ -909,6 +909,28 @@ func TestLeaderController_NotificationsCloseLeader(t *testing.T) {
 	assert.NoError(t, walFactory.Close())
 }
 
+func TestLeaderController_NotificationsWhenNotReady(t *testing.T) {
+	var shard int64 = 1
+
+	kvFactory, _ := kv.NewPebbleKVFactory(testKVOptions)
+	walFactory := newTestWalFactory(t)
+
+	lc, _ := NewLeaderController(Config{}, common.DefaultNamespace, shard, newMockRpcClient(), walFactory, kvFactory)
+	_, _ = lc.NewTerm(&proto.NewTermRequest{ShardId: shard, Term: 1})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	stream := newMockGetNotificationsServer(ctx)
+
+	// Get notification should fail if the leader controller is not fully initialized
+	err := lc.GetNotifications(&proto.NotificationsRequest{ShardId: shard}, stream)
+	assert.ErrorIs(t, err, context.Canceled)
+
+	assert.NoError(t, lc.Close())
+	assert.NoError(t, kvFactory.Close())
+	assert.NoError(t, walFactory.Close())
+}
+
 func TestLeaderController_List(t *testing.T) {
 	var shard int64 = 1
 

